### PR TITLE
Added model distillation guide

### PR DESCRIPTION
**Pull request by MichelBartels**
This adds a guide describing how and why to use model distillation. It also contains a table listing the performance improvements of model distillation for different models.

### CODE REVIEW

File: docs/latest/guides/model_distillation.mdx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+`distillation_loss_weight` specifies the weight given to the distillation loss in relation to the loss based on the label.
+Most times, you should set this to a relatively high value (e.g. 0.75).
+
+## What performance should I expect?
+
+Here are the results of some of our test and the parameters used:
+| Student model | Teacher model | Dataset | `temperature` | `distillation_loss_weight` | Distilled model EM | Distilled model F1 | Baseline EM | Baseline F1 | Teacher EM | Teacher F1 |
+|-|-|-|-|-|-|-|-|-|-|-|
+| prajjwal1/bert-medium | deepset/bert-large-uncased-whole-word-masking-squad2 | SQuADv2 | 5.0 | 1.0 | 68.6% | 72.8% | 65.6% | 69.6% | 78.9% | 83.2%7
```
Comment (julian-risch) [inline docs/latest/guides/model_distillation.mdx]: I see that the table already has many columns. However, one thing that's currently missing but important is the name (with a link to the model hub page) of the distilled model. Please add these names and links.

File: docs/latest/guides/model_distillation.mdx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+`temperature` specifies the certainty of the teacher model. Usually language models are very certain that one answer span is correct and assign all other answer spans a very low probability. This is not what we want for distillation as this would be very similar to just having a label. You can correct for this using a higher temperature.
+On the other hand, we also don't want the probabilities to be too similar as this also wouldn't create a meaningful training signal. For this, we can decrease the temperature.
+In most experiments, we found a temperature between 1 and 5 to be most useful.
+
+`distillation_loss_weight` specifies the weight given to the distillation loss in relation to the loss based on the label.
+Most times, you should set this to a relatively high value (e.g. 0.75).
+
+## What performance should I expect?
+
+Here are the results of some of our test and the parameters used:
```
Comment (julian-risch) [inline docs/latest/guides/model_distillation.mdx]: Maybe "some of our test" -> "some of our tests"? Maybe we can stress here something like: "We are actively working on extending model distillation features in haystack. The following table shows experiment results of the first version of model distillation in haystack. Stay tuned for updates!"

File: docs/latest/guides/model_distillation.mdx [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# Model distillation
```
Comment (julian-risch) [inline docs/latest/guides/model_distillation.mdx]: Let's capitalize this headline and also the other headlines as in the other guides on the documentation website. Here: "Model Distillation"

File: docs/latest/guides/model_distillation.mdx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Haystack now allows for logits-based distillation of large teacher models into smaller student models.
+
+## Why should I want to do distil a language model?
+
+In most cases, a larger model size leads to better performance. But the more parameters a model has, the more resource-intensive it is. This makes deployment more difficult and increases latency.
```
Comment (julian-risch) [inline docs/latest/guides/model_distillation.mdx]: Let's replace "performance" with "result quality" to disambiguate here. It's not about runtime, for example.

File: docs/latest/guides/model_distillation.mdx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+For this reason, choosing the right reader is always a trade-off between performance and deployment cost. Model distillation is one way to bridge the performance gap between a large language model (the teacher) and a small model (the student).
```
Comment (julian-risch) [inline docs/latest/guides/model_distillation.mdx]: Let's replace "performance" with "result quality" here too.

File: docs/latest/guides/model_distillation.mdx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+## What parameters should I use?
+
+Model distillation is controlled by two parameters: `temperature` and `distillation_loss_weight`.
+
+`temperature` specifies the certainty of the teacher model. Usually language models are very certain that one answer span is correct and assign all other answer spans a very low probability. This is not what we want for distillation as this would be very similar to just having a label. You can correct for this using a higher temperature.
+On the other hand, we also don't want the probabilities to be too similar as this also wouldn't create a meaningful training signal. For this, we can decrease the temperature.
+In most experiments, we found a temperature between 1 and 5 to be most useful.
+
+`distillation_loss_weight` specifies the weight given to the distillation loss in relation to the loss based on the label.
+Most times, you should set this to a relatively high value (e.g. 0.75).
```
Comment (julian-risch) [inline docs/latest/guides/model_distillation.mdx:58]: Please add here two sentences on what the extreme values mean (0 and 1). I think it helps users understanding the parameter, explains that we don't always need labels for model distillation and also thereby mentions the range of this parameter (between 0 and 1).

**Review by julian-risch — CHANGES_REQUESTED** [review]
Hi @MichelBartels the text looks already quite good to me. Just left some minor comments also to make sure that this guide is similar to the other ones we already have.
Regarding the table I am still a bit unsure what columns to include or whether to change rows with columns. Maybe we could also split it in two tables, for example, one with F1 score and one with exact match. I am afraid that the table is too large at the moment.

**Comment by julian-risch** [comment]
Maybe we could also add a sentence to explain that the current model distillation method that we are using aims to improve the result quality of small models. In contrast, pruning techniques would try to make an existing model smaller by keeping the same result quality. I liked your "bridging the gap" sentence here! 👍 Maybe we can elaborate a bit more there. Users might have heard about model distillation or distilled models somewhere else and it could be that they are expecting something else than what we provide.

**Comment by brandenchan** [comment]
Cool to see this in action! Thanks for your work on distillation and also on documenting it. Just want to let you know that to make this page show on the website you will need to add an entry into ` haystack-website/docs/latest/menu.json` - this file defines what will appear on the left Table of Contents.

File: docs/latest/guides/model_distillation.mdx (hunk 1 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx]: I have added the links now.

File: docs/latest/guides/model_distillation.mdx (hunk 2 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx]: Okay, that first one was a typo. I added the second part.

File: docs/latest/guides/model_distillation.mdx (hunk 3 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx]: Okay, I have changed that.

File: docs/latest/guides/model_distillation.mdx (hunk 4 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx]: I replaced it with quality of results and accuracy. I hope that's also okay.

File: docs/latest/guides/model_distillation.mdx (hunk 5 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx]: Addressed in answer above

File: docs/latest/guides/model_distillation.mdx (hunk 6 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx:58]: Okay, I have added a few sentences explaining that.

**Comment by MichelBartels** [comment]
I have now added a sentence referring to pruning. I also split the table into two. One for the parameters and one for the results.

**Comment by julian-risch** [comment]
Then what's left is mainly editing the `haystack-website/docs/latest/menu.json` file

File: docs/latest/guides/model_distillation.mdx (hunk 1 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx]: The first link is not yet working as I am still working on the model upload. I will write an update message later when it's finished.

**Comment by MichelBartels** [comment]
I have now updated the menu.json. However, I am not entirely certain I did this correctly. I have added an entry smilarly to the other guides, but haven't specified the filename. I am not sure where I have to add that.

File: docs/latest/guides/model_distillation.mdx (hunk 1 above)
Comment (MichelBartels) [reply docs/latest/guides/model_distillation.mdx]: All models are uploaded now.

**Comment by MichelBartels** [comment]
Currently, the guide isn't shown in the sidebar. Also, despite working in the Github preview the table isn't correctly rendered. I am not quite sure what is causing those issues

@brandenchan Could you help with this?

File: docs/latest/menu.json [hunk 7]
```diff
@@ -46,7 +46,8 @@
       {"slug": "annotation", "title": "Annotation Tool"},
       {"slug": "rest-api", "title": "REST API"},
       {"slug": "chatbots", "title": "Chatbot Integration"},
-      {"slug": "table-qa", "title": "Table Question Answering"}
+      {"slug": "table-qa", "title": "Table Question Answering"},
+      {"slug": "model-distillation", "title": "Model Distillation"},
```
Comment (julian-risch) [inline docs/latest/menu.json]: @MichelBartels The last comma here behind `"Model Distillation"}` could be a problem, I think.

Comment (MichelBartels) [reply docs/latest/menu.json]: I have now changed this line. This would make sense as trailing commas aren't part of the JSON specification.

**Review by julian-risch — APPROVED** [review]
Let's wait with merging until the preview in vercel looks good.
